### ISSUE TITLE
Terminal doesn't launch under 64 bit

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/vscode#27915]
- VSCode Version: 1.13
- OS Version: Windows 64 / Code 64

I have setup my terminal path like this: `"terminal.integrated.shell.windows": "C:\\WINDOWS\\Sysnative\\cmd.exe"`

Trying to open the terminal results in the blank terminal with a cursor.

**Comment by dbaeumer** [comment]
`"C:\\WINDOWS\\Sysnative\\bash.exe"` doesn't work either.

**Comment by dbaeumer** [comment]
Commenting out the setting which defaults to PowerShell works.

**Comment by chriswiley** [comment]
Using  VSCode Insiders 1.15 64-bit in Windows 10 and can't get bash to work either. Powershell works fine. But when I upgraded to 64-bit bash stopped working.

**Comment by dragonwolf83** [comment]
The 64-bit builds of VS Code need to update the default configuration paths that reference `sysnative` and change to `system32`.  The `sysnative` path is a virtualized path that only exists within 32-bit apps as an escape to find the real 64-bit directory which has always been `system32`.

So in your example, you need to use `"C:\\WINDOWS\\System32\\cmd.exe"` to get 64-bit cmd. 

If you need to get the 32-bit versions, you would need to use `SysWow64`.

**Comment by Tyriar** [comment]
Testing this out, here's what I found:

| terminal.integrated.shell.windows | Works?
|---|---
| C:\\WINDOWS\\System32\\WindowsPowerShell\\v1.0\\powershell.exe | ✔️
| C:\\WINDOWS\\Sysnative\\WindowsPowerShell\\v1.0\\powershell.exe | 🚫  
| C:\\WINDOWS\\SysWow64\\WindowsPowerShell\\v1.0\\powershell.exe | ✔️ (PSReadline not loaded)
| C:\\WINDOWS\\System32\\cmd.exe | ✔️ 
| C:\\WINDOWS\\Sysnative\\cmd.exe | 🚫 
| C:\\WINDOWS\\SysWow64\\cmd.exe | ✔️
| C:\\Program Files\\Git\\bin\\bash.exe | ✔️ 

This was caused by me not understanding properly how `Sysnative` works. Here's my new understanding:

- On 64-bit machine:
  - `SysWow64` only exists on 64-bit machines
  - In 64-bit application:
    - `SysWow64` contains 32-bit binaries
    - `System32` contains 64-bit binaries (nothing confusing here)
    - `Sysnative` doesn't exist
  - In 32-bit application:
    - Applications that request `System32` redirects to `SysWow64`
    - `Sysnative` only exists on 64-bit machines running 32-bit applications
    - `Sysnative` allows a 32-bit program to access the actual `System32` containing the 64-bit binaries (instead of redirecting)

Given this, here's the logic we'll go with when selecting `powershell.exe`, `cmd.exe` and `bash.exe` paths:

| Windows arch | Application arch | Path
|---|---|---
| 64 | 64 | `System32`
| 64 | 32 | `Sysnative`\*
| 32 | 64 | N/A
| 32 | 32 | `System32`

*\* perhaps we shouldn't bother with this and just use System32 (-> SysWow64) since the user explicitly requested the 32-bit version? It's probably best to assume they actually do want the 64-bit version as using the 32-bit one can lead to obscure messages that make VS Code look unreliable (like PSReadLine not working)*

**Comment by Tyriar** [comment]
We also need some way of checking if Windows is 64-bit inside a 32-bit process. Right now this is being done by checking for the `PROCESSOR_ARCHITEW6432` environment variable but I'm not sure this is correct as I don't see this environment variable on my machine, I do however see a `PROCESSOR_ARCHITECTURE` var (`"AMD64"`). This is the reason why the terminal works just fine using the default shell selections for me. I need to look into how to properly detect 32-bit app on a 64-bit machine.

**Comment by Tyriar** [comment]
Docs to the rescue! https://msdn.microsoft.com/en-us/library/aa384274.aspx

**Comment by Tyriar** [comment]
I pushed a fix for this. If someone who was experiencing the issue check with the defaults tomorrow that would be great. If it is happening, please let me know what the value for `terminal.integrated.shell.windows` is coming out as in the defaults. Also remember it's expected that if Sysnative is in your user settings it will not work.

**Comment by chriswiley** [comment]
I am running 1.15 right now and used:
```
"terminal.integrated.shell.windows": "C:\\WINDOWS\\System32\\bash.exe"
```
To get bash working. Is this the recommended path to use? I mean, it is working now.

**Comment by Tyriar** [comment]
@chriswiley yes that's the recommended path

**Comment by seanmcbreen** [comment]
@Tyriar I ran into this on my fresh 64bit Insiders.  I had CMD set up `sysnative` and I would get no terminal `undefined`.  I had to manually change this to the `System32` path and things worked fine.  The hover suggested I should do the opposite and intuitively `System32` seams like the wrong thing on my 64bit VS Code instance.

Either way I had to take a manual step to get things working it would be great to avoid that.

![image](https://user-images.githubusercontent.com/1051190/28584243-e108ee0c-7120-11e7-95a5-2b3fcfce0634.png)
